### PR TITLE
Merge from master into ort_training

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Additional to the merge, another commit was added to update existing ORT Training code after breaking changes from master. Refer to ffddf8e for further details

The following files conflicted and required manual changes, so pay special attention to them:

    onnxruntime/core/platform/env.h
    onnxruntime/core/platform/posix/env.cc
    onnxruntime/core/platform/windows/env.cc
    onnxruntime/core/providers/cuda/tensor/slice_impl.cu
    onnxruntime/core/session/inference_session.cc
    onnxruntime/python/session.py
    onnxruntime/test/python/onnx_backend_test_series.py
    onnxruntime/test/python/onnxruntime_test_python.py

Performance test:
Using LAMB optimizer
FP16, seq=128, bs=66, throughput=185.888
FP16, seq=512, bs=10, throughput=35.7677
FP32, seq=128, bs=33, throughput=41.7119
FP32, seq=512, bs=5, throughput=9.20151

e2e test:
https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=122361

Convergence test:
orange is baseline
image

### CODE REVIEW

File: onnxruntime/test/python/onnxruntime_test_python.py [hunk 1]
```diff
@@ -12,14 +12,25 @@
 
 class TestInferenceSession(unittest.TestCase):
 
+    def get_name(self, name):
```
Comment (edgchen1) [inline onnxruntime/test/python/onnxruntime_test_python.py]: perhaps we should consistently use self.get_name() or get_name() from helper. @liqunfu thoughts?

Comment (thiagocrepaldi) [reply onnxruntime/test/python/onnxruntime_test_python.py]: I talked to Liqun and he mentioned that the code that is ort_training is the correct one, so i updated the PR to only use get_name from helper.py and remove it from onnxruntime_test_python.py

**Comment by SherlockNoMad** [comment]
Please verify no regression in perf for both fp32 and fp16 (Please fill in the throughput numbers in the commit note)
Please verify no regression in convergence 
Please verify Onnxruntime_Linux_GPU_E2E_Test passes (this needs to be manually scheduled)

**Comment by SherlockNoMad** [comment]
I didn't see cuda/tensor/slice_impl.cu showing up here...
Is there no change?

**Comment by thiagocrepaldi** [comment]
> 
> 
> I didn't see cuda/tensor/slice_impl.cu showing up here...
> Is there no change?

All conflicts were resolved before commit, so they dont show up here. Check master version for details on this one: https://github.com/microsoft/onnxruntime/blob/c8f5e6e632251821300237812401dbe7fa2bcf1c/onnxruntime/core/providers/cuda/tensor/slice_impl.cu#L32

**Comment by edgchen1** [comment]
e2e test passed: https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=122361

**Pull request merged**